<a href="https://colab.research.google.com/github/anashenkonina-collab/python_for_ds_task/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22hw_10_3_%D0%A0%D0%BE%D0%B7%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D1%96_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BE%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B8_%D0%B4%D0%B0%D0%BD%D0%B8%D1%85_%D1%83_Pandas_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
data_path = '/content/drive/MyDrive/Data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [5]:
df.sort_values(by = 'Unit price', ascending = False).head(5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
983,148-41-7930,C,Naypyitaw,Normal,Male,Health and beauty,99.96,7,34.986,734.706,1/23/2019,10:33,Cash,699.72,4.761905,34.986,6.1
122,219-22-9386,B,Mandalay,Member,Male,Sports and travel,99.96,9,44.982,944.622,3/9/2019,17:26,Credit card,899.64,4.761905,44.982,4.2
930,641-62-7288,B,Mandalay,Normal,Male,Home and lifestyle,99.92,6,29.976,629.496,3/24/2019,13:33,Ewallet,599.52,4.761905,29.976,7.1
494,437-53-3084,B,Mandalay,Normal,Male,Fashion accessories,99.89,2,9.989,209.769,2/26/2019,11:48,Ewallet,199.78,4.761905,9.989,7.1
283,667-92-0055,A,Yangon,Member,Male,Health and beauty,99.83,6,29.949,628.929,3/4/2019,15:02,Ewallet,598.98,4.761905,29.949,8.5


2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [28]:
electronics = df[df['Product line'] == 'Electronic accessories']
unit_price_max = electronics[electronics['Unit price'] == electronics['Unit price'].max()]
quantity_max = electronics[electronics['Quantity'] == electronics['Quantity'].max()]
print(unit_price_max)
print(quantity_max)

      Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5  
      Invoice ID Branch       City Customer type  Gender  \
75   732-94-0499      C  Naypyitaw        Normal  Female   
105  704-48-3927      A     Yangon        Member    Male   
109  861-77-0145      C  Naypyitaw        Member    Male   
290  139-20-0155      B   Mandalay        Member    Male   
314  227-07-4446      C  Naypyitaw        Member  Female   
457  533-33-5337      B   Mandalay        Normal    Male   
458  325-90-8763      C  Naypyitaw        Member  Female   
481  246-11-3901      C  Naypyitaw        Normal  Female   

При перевірці бачимо, що не існує одного рядка у категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною, тому  виводимо рядок(інвойс) з максимальним значенням "Total"

In [23]:
electronics = df[df['Product line'] == 'Electronic accessories']
result = electronics[electronics['Total'] == electronics['Total'].max()]
print(result.T)

                                            209
Invoice ID                          817-69-8206
Branch                                        B
City                                   Mandalay
Customer type                            Normal
Gender                                   Female
Product line             Electronic accessories
Unit price                                99.73
Quantity                                      9
Tax 5%                                  44.8785
Total                                  942.4485
Date                                   3/2/2019
Time                                      19:42
Payment                             Credit card
cogs                                     897.57
gross margin percentage                4.761905
gross income                            44.8785
Rating                                      6.5


In [29]:
electronics.sort_values('Total', ascending=False).head(1)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
209,817-69-8206,B,Mandalay,Normal,Female,Electronic accessories,99.73,9,44.8785,942.4485,3/2/2019,19:42,Credit card,897.57,4.761905,44.8785,6.5


3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [59]:
def Rating_Category(rating):
  if rating >= 8:
    return 'High'
  if rating >= 6:
    return 'Medium'
  else:
    return 'Low'

df['Rating_Category'] =  df['Rating'].apply(Rating_Category)
df[['Rating','Rating_Category']]

,Rating,Rating_Category
0,9.1,High
1,9.6,High
2,7.4,Medium
3,8.4,High
4,5.3,Low
...,...,...
995,6.2,Medium
996,4.4,Low
997,7.7,Medium
998,4.1,Low


3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [62]:
df['Rating_Category'].value_counts(normalize=True)

,proportion
Rating_Category,
Medium,0.356
High,0.329
Low,0.315


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [68]:
time = '11:28'

In [73]:
def get_hour(time):
    return int(time.split(':')[0])

print(get_hour(time))



11


4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [77]:
df['Hour'] =  df['Time'].apply(get_hour)
df[['Time','Hour']]


,Time,Hour
0,13:08,13
1,10:29,10
2,13:23,13
3,20:33,20
4,10:37,10
...,...,...
995,13:46,13
996,17:16,17
997,13:22,13
998,15:33,15


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [79]:
df['Hour'].value_counts()

,count
Hour,
19,113
13,103
15,102
10,101
18,93
11,90
12,89
14,83
16,77


Найбільше покупок здійснюється о 19:00 (113 транзакцій) -кінець робочого дня, а також у 13:00  - обідній час

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [80]:
df.groupby('City')['Total'].mean()

,Total
City,
Mandalay,319.872506
Naypyitaw,337.099715
Yangon,312.354031


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [85]:
df.groupby('City')['Total'].agg(['count','mean', 'min', 'max'])

,count,mean,min,max
City,,,,
Mandalay,332,319.872506,18.6375,1022.49
Naypyitaw,328,337.099715,10.6785,1042.65
Yangon,340,312.354031,12.6945,1039.29


Рекомендація для мережі супермаркетів
Naypyitaw (найменше покупок — 328, але найвищий середній чек — 337.10 і максимальний чек — 1042.65): клієнти витрачають багато, але заходять рідко — варто стимулювати частоту відвідувань (програми лояльності, знижки на повторні покупки).
Широкий розкид чеків у Naypyitaw (10.68–1042.65, найширший серед міст) свідчить про неоднорідну аудиторію — варто сегментувати клієнтів для точніших пропозицій.

Yangon (найбільше покупок — 340, але найнижчий середній чек — 312.35): клієнти активні, але чек низький — варто підвищувати середню суму покупки.


7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [86]:
df.to_csv('supermarket_sales_enriched.csv', index=False)